src/reanimated2/hook/useEvent.ts

```diff
@@ -49,13 +48,18 @@ export function useEvent<Event extends object, Context = never>(
   eventNames: string[] = [],
   rebuild = false
 ): EventHandlerInternal<Event> {
-  const initRef = useRef<WorkletEventHandler<Event> | null>(null);
+  const initRef = useRef<EventHandlerInternal<Event>>(null!);
   if (initRef.current === null) {
-    initRef.current = new WorkletEventHandler<Event>(handler, eventNames);
+    const workletEventHandler = new WorkletEventHandler<Event>(
+      handler,
+      eventNames
+    );
+    initRef.current = { workletEventHandler };
   } else if (rebuild) {
-    initRef.current.updateWorklet(handler);
+    const workletEventHandler = initRef.current.workletEventHandler;
+    workletEventHandler.updateWorklet(handler);
+    initRef.current = { workletEventHandler };
```

Comment:
Is there any way that reassignment during render can be avoided? This still violates React's rules — it means a render is "observable" even if it has not been committed. In React, this can happen with newer features like Transitions. If you update state in a Transition, but this Transition has not completed yet, it should *not* reflect in the user interface. However, assigning a mutable value during rendering and then reading from that value means that the change from an uncommitted Transition render would reflect for any further re-renders (even before this Transition commits — or even if it never ends up being committed).

Reply:
@gaearon I think we could reassign with `useEffect` here, would that be good enough?

As a side note, I'm a bit confused with all these React features. From what you're saying it turns out that now you shouldn't assign to Refs during render - meaning that everything has to be pushed to `useEffect`. Is this intentional or am I missing some information here?

Reply:
It's not really new — in React, rendering is supposed to be pure. This means that the fact that something was rendered should be unobservable and shouldn't influence rendering output of other renders. 

Do these look helpful? https://react.dev/reference/rules

Reply:
Also here: https://react.dev/reference/react/useRef

![Screenshot 2024-05-29 at 00 08 39](https://github.com/software-mansion/react-native-reanimated/assets/810438/b2c6b1e8-f6bb-43a8-87dc-5fd283eae4c3)

Reply:
Damn, now I feel stupid 😅